src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader.cs

```diff
@@ -10,42 +10,36 @@ namespace System.Buffers.Reader
         private SequencePosition _currentPosition;
         private SequencePosition _nextPosition;
 
+        /// <summary>
+        /// Create a <see cref="BufferReader" over the given <see cref="ReadOnlySequence{byte}"/>./>
+        /// </summary>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private BufferReader(in ReadOnlySequence<byte> buffer)
+        public BufferReader(in ReadOnlySequence<byte> buffer)
         {
-            CurrentSegmentIndex = 0;
+            CurrentSpanIndex = 0;
             ConsumedBytes = 0;
             Sequence = buffer;
             _currentPosition = Sequence.Start;
             _nextPosition = _currentPosition;
 
-            if (buffer.TryGet(ref _nextPosition, out ReadOnlyMemory<byte> memory, true))
+            if (buffer.TryGet(ref _nextPosition, out ReadOnlyMemory<byte> memory, advance: true))
```

Comment:
Special case IsSingleSegment to avoid the unnecessary call to ReadOnlySequence\<byte\>.TryGet?

```C#
if (buffer.IsSingleSegment)
{
    CurrentSpan = buffer.First.Span;
    End = CurrentSpan.Length == 0;
}
else if (buffer.TryGet(...))
{
...
}
else
{
...
}
```

Reply:
Tanked multisegment- I'll look a bit more